File: python/dataset/dataset_utils.py
#renzi

"""Dataset extraction, preprocessing, validation, and evaluation utilities"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Sequence
import base64
import json
import random
import zlib

import cv2
import numpy as np


#paths

PYTHON_ROOT = Path(__file__).resolve().parents[1]
PROJECT_ROOT = Path(__file__).resolve().parents[2]
DATASET_ROOT = PROJECT_ROOT / "dataset"

BOUNDING_BOXES_TRAIN_ROOT = DATASET_ROOT / "fsoco_bounding_boxes_train"

SEGMENTATION_TRAIN_ROOT = DATASET_ROOT / "fsoco_segmentation_train"

SEGMENTATION_TEST_ROOT = DATASET_ROOT / "test_set" / "segmentation_test"
SEGMENTATION_TEST_ANNOTATION_DIR = SEGMENTATION_TEST_ROOT / "ann"
SEGMENTATION_TEST_IMAGE_DIR = SEGMENTATION_TEST_ROOT / "img"



BACKGROUND_ID = 0
YELLOW_CONE_ID = 1
BLUE_CONE_ID = 2
SMALL_ORANGE_CONE_ID = 3
BIG_ORANGE_CONE_ID = 4
IGNORE_ID = 255

NUM_CLASSES = 5

CONE_CLASS_IDS = (YELLOW_CONE_ID, BLUE_CONE_ID, SMALL_ORANGE_CONE_ID, BIG_ORANGE_CONE_ID)

CLASS_ID_TO_NAME = {
    BACKGROUND_ID: "background",
    YELLOW_CONE_ID: "yellow_cone",
    BLUE_CONE_ID: "blue_cone",
    SMALL_ORANGE_CONE_ID: "small_orange_cone",
    BIG_ORANGE_CONE_ID: "big_orange_cone",
}

# FSOCO uses both unprefixed detection labels and seg_* segmentation labels.
CLASS_NAME_TO_ID = {
    "yellow_cone": YELLOW_CONE_ID,
    "blue_cone": BLUE_CONE_ID,
    "orange_cone": SMALL_ORANGE_CONE_ID,
    "small_orange_cone": SMALL_ORANGE_CONE_ID,
    "large_orange_cone": BIG_ORANGE_CONE_ID,
    "big_orange_cone": BIG_ORANGE_CONE_ID,
    "unknown_cone": IGNORE_ID,
    "seg_yellow_cone": YELLOW_CONE_ID,
    "seg_blue_cone": BLUE_CONE_ID,
    "seg_orange_cone": SMALL_ORANGE_CONE_ID,
    "seg_small_orange_cone": SMALL_ORANGE_CONE_ID,
    "seg_large_orange_cone": BIG_ORANGE_CONE_ID,
    "seg_big_orange_cone": BIG_ORANGE_CONE_ID,
    "seg_unknown_cone": IGNORE_ID,
}

# Detection models use four zero-based classes; segmentation reserves 0 for background.
SEGMENTATION_ID_TO_DETECTION_ID = {
    YELLOW_CONE_ID: 0,
    BLUE_CONE_ID: 1,
    SMALL_ORANGE_CONE_ID: 2,
    BIG_ORANGE_CONE_ID: 3,
}

DETECTION_ID_TO_SEGMENTATION_ID = {detection_id: segmentation_id for segmentation_id, detection_id in SEGMENTATION_ID_TO_DETECTION_ID.items()}

DETECTION_ID_TO_NAME = { 0: "yellow_cone", 1: "blue_cone", 2: "small_orange_cone", 3: "big_orange_cone"}

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png"}

@dataclass(frozen=True)
class Box:
    """Bounding box using half-open xyxy coordinates: [x_min, x_max) x [y_min, y_max)"""

    x_min: int
    y_min: int
    x_max: int
    y_max: int
    class_id: int
    score: float | None = None

    def __post_init__(self) -> None:
        if self.x_min < 0 or self.y_min < 0:
            raise ValueError("Box minimum coordinates must be non-negative")
        if self.x_max <= self.x_min or self.y_max <= self.y_min:
            raise ValueError("Box maximum coordinates must be greater than minimum coordinates")
        if self.class_id not in CONE_CLASS_IDS:
            raise ValueError(f"Invalid shared cone class id: {self.class_id}")
        if self.score is not None and not 0.0 <= self.score <= 1.0:
            raise ValueError("Box score must be in the range [0, 1]")

    @property
    def area(self) -> int:
        return (self.x_max - self.x_min) * (self.y_max - self.y_min)

@dataclass(frozen=True)
class SegmentationInstance:
    """One cone instance with its class, bounding box, and binary mask"""

    class_id: int
    bbox: Box
    mask: np.ndarray

def class_name_to_id(class_name: str) -> int:
    """Convert an fsoco class name to the shared segmentation class id"""

    if not isinstance(class_name, str) or not class_name.strip():
        raise ValueError("class_name must be a non-empty string")

    normalized_name = class_name.strip().lower()

    if normalized_name not in CLASS_NAME_TO_ID:
        raise ValueError(f"Unknown fsoco class: {class_name}")

    return CLASS_NAME_TO_ID[normalized_name]


def segmentation_id_to_detection_id(class_id: int) -> int:
    """Convert a shared segmentation class id into a zero-based detection id"""

    if class_id not in SEGMENTATION_ID_TO_DETECTION_ID:
        raise ValueError(f"Class id {class_id} cannot be used as a detection class")

    return SEGMENTATION_ID_TO_DETECTION_ID[class_id]


def detection_id_to_segmentation_id(class_id: int) -> int:
    """Convert a zero-based detection id into the shared segmentation class id"""

    if class_id not in DETECTION_ID_TO_SEGMENTATION_ID:
        raise ValueError(f"Invalid detection class id: {class_id}")

    return DETECTION_ID_TO_SEGMENTATION_ID[class_id]


def prediction_to_box(bbox_xyxy: Sequence[float], detection_class_id: int, score: float) -> Box:
    """Convert one model prediction to the shared Box representation"""

    if len(bbox_xyxy) != 4:
        raise ValueError("bbox_xyxy must contain exactly four coordinates")

    x_min, y_min, x_max, y_max = (int(round(float(value))) for value in bbox_xyxy)
    return Box(x_min, y_min, x_max, y_max, detection_id_to_segmentation_id(int(detection_class_id)), float(score))

def ground_truth_to_box(bbox_xyxy, detection_class_id):
    x_min, y_min, x_max, y_max = (int(round(float(value))) for value in bbox_xyxy)

    return Box(x_min, y_min, x_max, y_max,
        detection_id_to_segmentation_id(
            int(detection_class_id)
        ),
        None,
    )

def load_annotation(annotation_path: str | Path) -> dict[str, Any]:
    """Load and validate the top-level structure of a json annotation"""

    annotation_path = Path(annotation_path)

    if not annotation_path.is_file():
        raise FileNotFoundError(f"Annotation file not found: {annotation_path}")

    try:
        with annotation_path.open("r", encoding="utf-8") as file:
            annotation = json.load(file)
    except json.JSONDecodeError as error:
        raise ValueError(f"Invalid json annotation: {annotation_path}") from error

    if not isinstance(annotation, dict):
        raise ValueError(f"Annotation root must be a json object: {annotation_path}")
    if not isinstance(annotation.get("objects", []), list):
        raise ValueError(f"Annotation 'objects' must be a list: {annotation_path}")

    return annotation


def load_image(image_path: str | Path) -> np.ndarray:
    """Load an image in opencv bgr format"""

    image_path = Path(image_path)

    if not image_path.is_file():
        raise FileNotFoundError(f"Image file not found: {image_path}")

    image = cv2.imread(str(image_path), cv2.IMREAD_COLOR)

    if image is None:
        raise RuntimeError(f"Cannot decode image: {image_path}")

    return image




def find_dataset_pairs(image_dir: str | Path, annotation_dir: str | Path) -> list[tuple[Path, Path]]:
    """Find deterministic image/annotation pairs inside one fsoco subset"""

    image_dir = Path(image_dir)
    annotation_dir = Path(annotation_dir)

    if not image_dir.is_dir():
        raise FileNotFoundError(f"Image directory not found: {image_dir}")
    if not annotation_dir.is_dir():
        raise FileNotFoundError(f"Annotation directory not found: {annotation_dir}")

    images_by_name: dict[str, Path] = {}
    images_by_stem: dict[str, list[Path]] = {}

    for image_path in sorted(image_dir.iterdir()):
        if not image_path.is_file() or image_path.suffix.lower() not in IMAGE_EXTENSIONS:
            continue
        images_by_name[image_path.name] = image_path
        images_by_stem.setdefault(image_path.stem, []).append(image_path)

    pairs: list[tuple[Path, Path]] = []
    missing_images: list[str] = []

    for annotation_path in sorted(annotation_dir.glob("*.json")):
        expected_image_name = annotation_path.name.removesuffix(".json")
        image_path = images_by_name.get(expected_image_name)

        if image_path is None:
            stem_candidates = images_by_stem.get(annotation_path.stem, [])
            if len(stem_candidates) > 1:
                raise ValueError(f"Ambiguous image stem for annotation {annotation_path.name}: {stem_candidates}")
            if len(stem_candidates) == 1:
                image_path = stem_candidates[0]

        if image_path is None:
            missing_images.append(annotation_path.name)
            continue

        pairs.append((image_path, annotation_path))

    if missing_images:
        preview = ", ".join(missing_images[:5])
        suffix = "" if len(missing_images) <= 5 else f" and {len(missing_images) - 5} more"
        raise FileNotFoundError(f"Missing images for annotations in {annotation_dir}: {preview}{suffix}")

    return pairs


def find_all_dataset_pairs(dataset_root: str | Path) -> list[tuple[Path, Path]]:
    """Find pairs in a direct ann/img root or in all one-level subset roots"""

    dataset_root = Path(dataset_root)

    if not dataset_root.is_dir():
        raise FileNotFoundError(f"Dataset root not found: {dataset_root}")

    direct_image_dir = dataset_root / "img"
    direct_annotation_dir = dataset_root / "ann"

    if direct_image_dir.is_dir() or direct_annotation_dir.is_dir():
        if not direct_image_dir.is_dir() or not direct_annotation_dir.is_dir():
            raise FileNotFoundError(f"Both img and ann directories are required in {dataset_root}")
        return find_dataset_pairs(direct_image_dir, direct_annotation_dir)

    all_pairs: list[tuple[Path, Path]] = []

    for subset_dir in sorted(path for path in dataset_root.iterdir() if path.is_dir()):
        image_dir = subset_dir / "img"
        annotation_dir = subset_dir / "ann"
        if image_dir.is_dir() or annotation_dir.is_dir():
            if not image_dir.is_dir() or not annotation_dir.is_dir():
                raise FileNotFoundError(f"Both img and ann directories are required in {subset_dir}")
            all_pairs.extend(find_dataset_pairs(image_dir, annotation_dir))

    if not all_pairs:
        raise ValueError(f"No image/annotation pairs found in {dataset_root}")

    return sorted(all_pairs, key=lambda pair: str(pair[0]))


def get_bounding_boxes_train_pairs() -> list[tuple[Path, Path]]:
    """Return pairs from dataset/fsoco_bounding_boxes_train"""
    return find_all_dataset_pairs(BOUNDING_BOXES_TRAIN_ROOT)


def get_segmentation_train_pairs() -> list[tuple[Path, Path]]:
    """Return pairs from dataset/fsoco_segmentation_train"""
    return find_all_dataset_pairs(SEGMENTATION_TRAIN_ROOT)


def train_validation_split(pairs: Sequence[tuple[Path, Path]], validation_fraction: float = 0.20, seed: int = 42) -> tuple[list[tuple[Path, Path]], list[tuple[Path, Path]]]:
    """Create a deterministic train/validation split without mutating the input"""

    if not 0.0 <= validation_fraction < 1.0:
        raise ValueError("validation_fraction must be in the range [0.0, 1.0)")

    shuffled_pairs = list(pairs)
    random.Random(seed).shuffle(shuffled_pairs)
    validation_size = round(len(shuffled_pairs) * validation_fraction)

    if len(shuffled_pairs) > 1 and validation_fraction > 0.0:
        validation_size = min(max(1, validation_size), len(shuffled_pairs) - 1)

    return shuffled_pairs[validation_size:], shuffled_pairs[:validation_size]


def decode_bitmap(bitmap_data: str) -> np.ndarray:
    """Decode a compressed Supervisely bitmap into a local boolean mask"""

    if not isinstance(bitmap_data, str) or not bitmap_data:
        raise ValueError("bitmap_data must be a non-empty base64 string")

    try:
        compressed_data = base64.b64decode(bitmap_data, validate=True)
        image_data = zlib.decompress(compressed_data)
    except (ValueError, zlib.error) as error:
        raise ValueError("Invalid compressed Supervisely bitmap data") from error

    decoded_image = cv2.imdecode(np.frombuffer(image_data, dtype=np.uint8), cv2.IMREAD_UNCHANGED)

    if decoded_image is None:
        raise ValueError("Could not decode Supervisely bitmap image")
    if decoded_image.ndim == 3 and decoded_image.shape[2] == 4:
        return decoded_image[:, :, 3] > 0
    if decoded_image.ndim == 2:
        return decoded_image > 0

    raise ValueError(f"Unexpected decoded bitmap shape: {decoded_image.shape}")


def bitmap_object_to_full_mask(obj: dict[str, Any], image_height: int, image_width: int) -> np.ndarray:
    """Convert one Supervisely bitmap object into a full-image boolean mask"""

    _validate_image_size(image_height, image_width)

    if obj.get("geometryType") != "bitmap":
        raise ValueError("The annotation object is not a bitmap")

    bitmap = obj.get("bitmap")

    if not isinstance(bitmap, dict):
        raise ValueError("Missing bitmap data in annotation object")

    bitmap_data = bitmap.get("data")
    origin = bitmap.get("origin")

    if bitmap_data is None or not isinstance(origin, (list, tuple)) or len(origin) != 2:
        raise ValueError("Bitmap annotation is missing valid data or origin")

    local_mask = decode_bitmap(bitmap_data)
    origin_x, origin_y = int(origin[0]), int(origin[1])
    object_height, object_width = local_mask.shape

    x_start = max(0, origin_x)
    y_start = max(0, origin_y)
    x_end = min(image_width, origin_x + object_width)
    y_end = min(image_height, origin_y + object_height)

    full_mask = np.zeros((image_height, image_width), dtype=bool)

    if x_start >= x_end or y_start >= y_end:
        return full_mask

    bitmap_x_start = x_start - origin_x
    bitmap_y_start = y_start - origin_y
    bitmap_x_end = bitmap_x_start + (x_end - x_start)
    bitmap_y_end = bitmap_y_start + (y_end - y_start)
    full_mask[y_start:y_end, x_start:x_end] = local_mask[bitmap_y_start:bitmap_y_end, bitmap_x_start:bitmap_x_end]

    return full_mask



def binary_mask_to_bbox(binary_mask: np.ndarray) -> tuple[int, int, int, int] | None:
    """Return the tight half-open xyxy box of a binary mask"""

    if binary_mask.ndim != 2:
        raise ValueError("binary_mask must be a 2D array")

    y_coordinates, x_coordinates = np.nonzero(binary_mask)

    if x_coordinates.size == 0:
        return None

    return int(x_coordinates.min()), int(y_coordinates.min()), int(x_coordinates.max()) + 1, int(y_coordinates.max()) + 1


def rectangle_object_to_bbox(obj: dict[str, Any], image_height: int, image_width: int) -> tuple[int, int, int, int] | None:
    """Convert one Supervisely rectangle into a clipped half-open xyxy box"""

    _validate_image_size(image_height, image_width)

    if obj.get("geometryType") != "rectangle":
        raise ValueError("The annotation object is not a rectangle")

    exterior = obj.get("points", {}).get("exterior", [])

    if not isinstance(exterior, list) or len(exterior) < 2:
        return None

    try:
        x_values = [float(point[0]) for point in exterior]
        y_values = [float(point[1]) for point in exterior]
    except (TypeError, ValueError, IndexError) as error:
        raise ValueError("Invalid rectangle exterior coordinates") from error

    x_min = max(0, int(np.floor(min(x_values))))
    y_min = max(0, int(np.floor(min(y_values))))
    x_max = min(image_width, int(np.floor(max(x_values))) + 1)
    y_max = min(image_height, int(np.floor(max(y_values))) + 1)

    if x_max <= x_min or y_max <= y_min:
        return None

    return x_min, y_min, x_max, y_max


def object_to_bbox(obj: dict[str, Any], image_height: int, image_width: int) -> tuple[int, int, int, int] | None:
    """Convert one bitmap or rectangle annotation object to a half-open xyxy box"""

    geometry_type = obj.get("geometryType")

    if geometry_type == "bitmap":
        return binary_mask_to_bbox(bitmap_object_to_full_mask(obj, image_height, image_width))
    if geometry_type == "rectangle":
        return rectangle_object_to_bbox(obj, image_height, image_width)

    return None


def annotation_to_instances(annotation_path: str | Path, image_height: int, image_width: int, bitmap_only: bool = False) -> list[Box]:
    """Extract one shared-class Box for every known cone instance"""

    _validate_image_size(image_height, image_width)
    annotation = load_annotation(annotation_path)
    instances: list[Box] = []

    for obj in annotation.get("objects", []):
        if not isinstance(obj, dict):
            continue
        if bitmap_only and obj.get("geometryType") != "bitmap":
            continue

        class_name = obj.get("classTitle")

        if class_name is None:
            continue

        class_id = class_name_to_id(class_name)

        if class_id == IGNORE_ID:
            continue

        coordinates = object_to_bbox(obj, image_height, image_width)

        if coordinates is not None:
            instances.append(Box(*coordinates, class_id))

    return instances


def annotation_to_segmentation_instances(annotation_path: str | Path, image_height: int, image_width: int) -> list[SegmentationInstance]:
    """Extract cone instances with their individual binary masks and bounding boxes"""

    _validate_image_size(image_height, image_width)
    annotation = load_annotation(annotation_path)
    instances: list[SegmentationInstance] = []

    for obj in annotation.get("objects", []):
        if not isinstance(obj, dict):
            continue

        if obj.get("geometryType") != "bitmap":
            continue

        class_name = obj.get("classTitle")

        if class_name is None:
            continue

        class_id = class_name_to_id(class_name)

        if class_id == IGNORE_ID:
            continue

        binary_mask = bitmap_object_to_full_mask(obj, image_height, image_width)
        coordinates = binary_mask_to_bbox(binary_mask)

        if coordinates is None:
            continue

        bbox = Box(*coordinates, class_id)
        instances.append(SegmentationInstance(class_id=class_id, bbox=bbox, mask=binary_mask))

    return instances

#validation helpers

def validate_dataset_pairs(pairs: Sequence[tuple[Path, Path]]) -> None:
    """Check that every image and JSON annotation can be loaded"""

    for image_path, annotation_path in pairs:
        load_image(image_path)
        load_annotation(annotation_path)


def _validate_image_size(image_height: int, image_width: int) -> None:
    if not isinstance(image_height, int) or not isinstance(image_width, int):
        raise TypeError("image_height and image_width must be integers")
    if image_height <= 0 or image_width <= 0:
        raise ValueError("image_height and image_width must be positive")

File: src/dataset_utils.cpp
//Renzi

#include "dataset_utils.h"

#include <algorithm>
#include <array>
#include <cctype>
#include <filesystem>
#include <map>
#include <optional>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>
#include <fstream>
#include <sstream>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <zlib.h>

namespace {

// Checks that image dimensions can be safely used to allocate masks.
void validate_image_size(int image_height, int image_width){
    if (image_height <= 0 || image_width <= 0) {
        throw std::invalid_argument("Image dimensions must be positive.");
    }
}

// Validates the basic format expected for semantic masks.
void validate_mask(const cv::Mat& mask){
    if (mask.empty() || mask.dims != 2 || mask.channels() != 1) {
        throw std::invalid_argument("Mask must be a non-empty single-channel image.");
    }
}

// Returns true for the image formats accepted by the dataset loader.
bool is_supported_image(const std::filesystem::path& path){
    std::string extension = path.extension().string();
    std::transform(extension.begin(), extension.end(), extension.begin(), [](unsigned char character) { return static_cast<char>(std::tolower(character)); });
    return extension == ".jpg" || extension == ".jpeg" || extension == ".png" || extension == ".bmp";
}

// Converts fsoco class labels into the internal numeric identifiers.
int class_name_to_id(const std::string& class_name){
    static const std::unordered_map<std::string, int> class_ids = {
        {"yellow_cone", YELLOW_CONE_ID},
        {"blue_cone", BLUE_CONE_ID},
        {"orange_cone", SMALL_ORANGE_CONE_ID},
        {"small_orange_cone", SMALL_ORANGE_CONE_ID},
        {"large_orange_cone", BIG_ORANGE_CONE_ID},
        {"big_orange_cone", BIG_ORANGE_CONE_ID},
        {"unknown_cone", IGNORE_ID},
        {"seg_yellow_cone", YELLOW_CONE_ID},
        {"seg_blue_cone", BLUE_CONE_ID},
        {"seg_orange_cone", SMALL_ORANGE_CONE_ID},
        {"seg_small_orange_cone", SMALL_ORANGE_CONE_ID},
        {"seg_large_orange_cone", BIG_ORANGE_CONE_ID},
        {"seg_big_orange_cone", BIG_ORANGE_CONE_ID},
        {"seg_unknown_cone", IGNORE_ID}
    };

    std::string normalized_name = class_name;
    std::transform(normalized_name.begin(), normalized_name.end(), normalized_name.begin(), [](unsigned char character) { return static_cast<char>(std::tolower(character)); });

    const auto iterator = class_ids.find(normalized_name);

    if (iterator == class_ids.end()){
        throw std::invalid_argument("Unknown FSOCO class: " + class_name);
    }

    return iterator->second;
}

    // Replaces json null values outside strings so OpenCV FileStorage can parse the file.
    std::string replace_json_nulls(const std::string& json_text){
        std::string result;
        bool inside_string = false;
        bool escaped = false;

        for (std::size_t index = 0; index < json_text.size(); ++index){
            const char character = json_text[index];

            if (escaped){
                result += character;
                escaped = false;
                continue;
            }

            if (character == '\\' && inside_string){
                result += character;
                escaped = true;
                continue;
            }

            if (character == '"') {
                inside_string = !inside_string;
                result += character;
                continue;
            }

            if (!inside_string && json_text.compare(index, 4, "null") == 0){
                result += "0";
                index += 3;
                continue;
            }

            result += character;
        }

        return result;
    }


    // Reads and parses a json annotation from disk.
    cv::FileStorage open_annotation(const std::filesystem::path& annotation_path){
        if (!std::filesystem::is_regular_file(annotation_path)){
            throw std::runtime_error("Annotation file not found: " + annotation_path.string());
        }

        std::ifstream file(annotation_path);

        if (!file.is_open()){
            throw std::runtime_error("Could not open annotation: " + annotation_path.string());
        }

        std::stringstream buffer;
        buffer << file.rdbuf();

        const std::string json_text = replace_json_nulls(buffer.str());

        cv::FileStorage annotation(json_text, cv::FileStorage::READ | cv::FileStorage::MEMORY | cv::FileStorage::FORMAT_JSON);

        if (!annotation.isOpened()){
            throw std::runtime_error("Could not parse annotation: " + annotation_path.string());
        }

        return annotation;
    }

// Decodes the base64 payload used by bitmap annotations.
std::vector<unsigned char> decode_base64(const std::string& input){
    static const std::string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    std::array<int, 256> lookup;
    lookup.fill(-1);

    for (std::size_t index = 0; index < alphabet.size(); ++index){
        lookup[static_cast<unsigned char>(alphabet[index])] = static_cast<int>(index);
    }

    std::vector<unsigned char> output;
    int value = 0;
    int bits = -8;

    for (unsigned char character : input){
        if (character == '='){
            break;
        }

        if (lookup[character] < 0){
            throw std::runtime_error("Invalid Base64 bitmap data.");
        }

        value = (value << 6) + lookup[character];
        bits += 6;

        if (bits >= 0) {
            output.push_back(static_cast<unsigned char>((value >> bits) & 0xFF));
            bits -= 8;
        }
    }

    return output;
}

// Decompresses the zlib-compressed bitmap payload.
std::vector<unsigned char> decompress_zlib(const std::vector<unsigned char>& compressed_data){
    z_stream stream{};
    stream.next_in = const_cast<Bytef*>(reinterpret_cast<const Bytef*>(compressed_data.data()));
    stream.avail_in = static_cast<uInt>(compressed_data.size());

    if (inflateInit(&stream) != Z_OK){
        throw std::runtime_error("Could not initialize zlib.");
    }

    std::vector<unsigned char> output;
    std::array<unsigned char, 16384> buffer{};
    int status = Z_OK;

    while (status != Z_STREAM_END){
        stream.next_out = buffer.data();
        stream.avail_out = static_cast<uInt>(buffer.size());
        status = inflate(&stream, Z_NO_FLUSH);

        if (status != Z_OK && status != Z_STREAM_END){
            inflateEnd(&stream);
            throw std::runtime_error("Could not decompress bitmap data.");
        }

        const std::size_t produced = buffer.size() - stream.avail_out;
        output.insert(output.end(), buffer.begin(), buffer.begin() + static_cast<std::ptrdiff_t>(produced));
    }

    inflateEnd(&stream);
    return output;
}

// Decodes a bitmap annotation into an OpenCV mask.
cv::Mat decode_bitmap(const std::string& bitmap_data){
    const std::vector<unsigned char> compressed_data = decode_base64(bitmap_data);
    const std::vector<unsigned char> image_data = decompress_zlib(compressed_data);

    if (image_data.empty()){
        throw std::runtime_error("Decoded bitmap data is empty.");
    }

    const cv::Mat encoded_image(1, static_cast<int>(image_data.size()), CV_8U, const_cast<unsigned char*>(image_data.data()));
    const cv::Mat decoded_image = cv::imdecode(encoded_image, cv::IMREAD_UNCHANGED);

    if (decoded_image.empty()){
        throw std::runtime_error("Could not decode bitmap image.");
    }

    cv::Mat source_mask;

    if (decoded_image.channels() == 4){
        cv::extractChannel(decoded_image, source_mask, 3);
    } else if (decoded_image.channels() == 1){
        source_mask = decoded_image;
    } else{
        throw std::runtime_error("Unexpected bitmap format.");
    }

    cv::Mat binary_mask;
    cv::compare(source_mask, 0, binary_mask, cv::CMP_GT);
    binary_mask /= 255;
    return binary_mask;
}

// Places an object-local bitmap into a full-size image mask.
cv::Mat object_to_full_mask(const cv::FileNode& object, int image_height, int image_width){
    const cv::FileNode bitmap = object["bitmap"];
    const cv::FileNode origin = bitmap["origin"];

    if (bitmap.empty() || bitmap["data"].empty() || origin.empty() || origin.size() != 2){
        throw std::runtime_error("Invalid bitmap annotation.");
    }

    const cv::Mat local_mask = decode_bitmap(static_cast<std::string>(bitmap["data"]));
    const int origin_x = static_cast<int>(origin[0]);
    const int origin_y = static_cast<int>(origin[1]);

    const int x_start = std::max(0, origin_x);
    const int y_start = std::max(0, origin_y);
    const int x_end = std::min(image_width, origin_x + local_mask.cols);
    const int y_end = std::min(image_height, origin_y + local_mask.rows);

    cv::Mat full_mask = cv::Mat::zeros(image_height, image_width, CV_8UC1);

    if (x_start >= x_end || y_start >= y_end){
        return full_mask;
    }

    const cv::Rect source_roi(x_start - origin_x, y_start - origin_y, x_end - x_start, y_end - y_start);
    const cv::Rect destination_roi(x_start, y_start, x_end - x_start, y_end - y_start);
    local_mask(source_roi).copyTo(full_mask(destination_roi));

    return full_mask;
}

// Extracts the tight bounding box of the non-zero mask pixels.
std::optional<std::array<int, 4>> mask_to_bbox(const cv::Mat& mask){
    std::vector<cv::Point> points;
    cv::findNonZero(mask, points);

    if (points.empty()){
        return std::nullopt;
    }

    const cv::Rect rectangle = cv::boundingRect(points);
    return std::array<int, 4>{rectangle.x, rectangle.y, rectangle.x + rectangle.width, rectangle.y + rectangle.height};
}

// Converts class IDs into BGR colors for visualization.
cv::Mat colorize_mask(const cv::Mat& mask){
    validate_mask(mask);
    cv::Mat color_mask(mask.rows, mask.cols, CV_8UC3);

    for (int row = 0; row < mask.rows; ++row){
        for (int column = 0; column < mask.cols; ++column){
            const int class_id = static_cast<int>(mask.at<unsigned char>(row, column));
            const cv::Scalar color = class_color_bgr(class_id);
            color_mask.at<cv::Vec3b>(row, column) = cv::Vec3b(static_cast<unsigned char>(color[0]), static_cast<unsigned char>(color[1]), static_cast<unsigned char>(color[2]));
        }
    }

    return color_mask;
}

}

// Checks whether an identifier belongs to one of the cone classes.
bool is_cone_class_id(int class_id){
    return class_id >= YELLOW_CONE_ID && class_id <= BIG_ORANGE_CONE_ID;
}

// Converts an internal class ID into a readable class name.
std::string class_id_to_name(int class_id){
    switch (class_id) {
        case BACKGROUND_ID:
            return "background";
        case YELLOW_CONE_ID:
            return "yellow_cone";
        case BLUE_CONE_ID:
            return "blue_cone";
        case SMALL_ORANGE_CONE_ID:
            return "small_orange_cone";
        case BIG_ORANGE_CONE_ID:
            return "big_orange_cone";
        case IGNORE_ID:
            return "ignore";
        default:
            throw std::invalid_argument("Invalid class ID: " + std::to_string(class_id));
    }
}

// Returns the BGR color associated with a class.
cv::Scalar class_color_bgr(int class_id){
    switch (class_id){
        case BACKGROUND_ID:
            return cv::Scalar(0, 0, 0);
        case YELLOW_CONE_ID:
            return cv::Scalar(0, 255, 255);
        case BLUE_CONE_ID:
            return cv::Scalar(255, 0, 0);
        case SMALL_ORANGE_CONE_ID:
            return cv::Scalar(0, 165, 255);
        case BIG_ORANGE_CONE_ID:
            return cv::Scalar(255, 0, 255);
        case IGNORE_ID:
            return cv::Scalar(255, 255, 255);
        default:
            throw std::invalid_argument("Invalid class ID: " + std::to_string(class_id));
    }
}

// Loads a color image and validates that decoding succeeded.
cv::Mat load_image(const std::filesystem::path& image_path){
    if (!std::filesystem::is_regular_file(image_path)){
        throw std::runtime_error("Image file not found: " + image_path.string());
    }

    const cv::Mat image = cv::imread(image_path.string(), cv::IMREAD_COLOR);

    if (image.empty()){
        throw std::runtime_error("Could not decode image: " + image_path.string());
    }

    return image;
}

// Matches test images and annotations by filename stem.
std::vector<DatasetPair> find_test_pairs(const std::filesystem::path& image_dir, const std::filesystem::path& annotation_dir){
    if (!std::filesystem::is_directory(image_dir) || !std::filesystem::is_directory(annotation_dir)){
        throw std::runtime_error("Test image or annotation directory not found.");
    }

    std::map<std::string, std::filesystem::path> images;

    for (const std::filesystem::directory_entry& entry : std::filesystem::directory_iterator(image_dir)){
        if (entry.is_regular_file() && is_supported_image(entry.path())){
            images[entry.path().filename().string()] = entry.path();
        }
    }

    std::vector<DatasetPair> pairs;

    for (const std::filesystem::directory_entry& entry : std::filesystem::directory_iterator(annotation_dir)){
        if (!entry.is_regular_file() || entry.path().extension() != ".json"){
            continue;
        }

        std::string image_name = entry.path().filename().string();
        image_name.erase(image_name.size() - 5);

        const auto image_iterator = images.find(image_name);

        if (image_iterator == images.end()){
            throw std::runtime_error("Missing image for annotation: " + entry.path().filename().string());
        }

        pairs.push_back({image_iterator->second, entry.path()});
    }

    std::sort(pairs.begin(), pairs.end(), [](const DatasetPair& first, const DatasetPair& second) { return first.image_path.string() < second.image_path.string(); });

    if (pairs.empty()){
        throw std::runtime_error("No test image/annotation pairs found.");
    }

    return pairs;
}

// Builds the semantic mask and bounding boxes from one fsoco annotation.
GroundTruth load_ground_truth(const std::filesystem::path& annotation_path, int image_height, int image_width){
    validate_image_size(image_height, image_width);

    cv::FileStorage annotation = open_annotation(annotation_path);
    const cv::FileNode objects = annotation["objects"];

    GroundTruth ground_truth;
    ground_truth.semantic_mask = cv::Mat::zeros(image_height, image_width, CV_8UC1);

    if (objects.empty()){
        return ground_truth;
    }

    for (const cv::FileNode& object : objects){
        if (static_cast<std::string>(object["geometryType"]) != "bitmap" || object["classTitle"].empty()) {
            continue;
        }

        const int class_id = class_name_to_id(static_cast<std::string>(object["classTitle"]));
        const cv::Mat object_mask = object_to_full_mask(object, image_height, image_width);
        ground_truth.semantic_mask.setTo(class_id, object_mask);

        if (class_id == IGNORE_ID){
            continue;
        }

        const std::optional<std::array<int, 4>> coordinates = mask_to_bbox(object_mask);

        if (!coordinates.has_value()){
            continue;
        }

        Box box;
        box.x_min = coordinates.value()[0];
        box.y_min = coordinates.value()[1];
        box.x_max = coordinates.value()[2];
        box.y_max = coordinates.value()[3];
        box.class_id = class_id;
        box.score = std::nullopt;
        ground_truth.boxes.push_back(box);
    }

    return ground_truth;
}

// Blends the colored semantic mask with the original image.
cv::Mat create_overlay(const cv::Mat& image_bgr, const cv::Mat& mask, double alpha){
    if (image_bgr.empty() || image_bgr.channels() != 3){
        throw std::invalid_argument("Image must be a non-empty three-channel image.");
    }

    validate_mask(mask);

    if (image_bgr.size() != mask.size()){
        throw std::invalid_argument("Image and mask sizes must match.");
    }

    if (alpha < 0.0 || alpha > 1.0){
        throw std::invalid_argument("Alpha must be in the range [0, 1].");
    }

    const cv::Mat color_mask = colorize_mask(mask);

    cv::Mat blended;
    cv::addWeighted(image_bgr, 1.0 - alpha, color_mask, alpha, 0.0, blended);

    cv::Mat foreground;
    cv::compare(mask, BACKGROUND_ID, foreground, cv::CMP_NE);

    cv::Mat overlay = image_bgr.clone();
    blended.copyTo(overlay, foreground);
    return overlay;
}
